pkg/storegateway/series_chunks.go

```diff
@@ -417,6 +422,183 @@ func (c *loadingSeriesChunksSetIterator) Err() error {
+		dst := s.parsedChunks[parsedChunksCount : parsedChunksCount+len(r.refs)]
+		ok, err := s.populateRange(rIdx, dst)
+		if err != nil {
+			return nil, fmt.Errorf("parsing chunk range (block %s, first ref %d, num chunks %d): %w", r.blockID, r.firstRef(), len(r.refs), err)
+		}
+		if !ok {
+			underfetchedRanges = append(underfetchedRanges, underfetchedChunksRangeIdx{
+				blockID:  r.blockID,
+				rangeIdx: rIdx,
+				parsed:   dst,
+			})
+		}
+
+		parsedChunksCount += len(r.refs)
+	}
+	return underfetchedRanges, nil
+}
+
+//nolint:unused // dead code while we are working on PR 3968
+func convertChunkEncoding(storageEncoding chunkenc.Encoding) (storepb.Chunk_Encoding, bool) {
+	switch storageEncoding {
+	case chunkenc.EncXOR:
+		return storepb.Chunk_XOR, true
+	case chunkenc.EncHistogram:
+		return storepb.Chunk_Histogram, true
+	case chunkenc.EncFloatHistogram:
+		return storepb.Chunk_FloatHistogram, true
+	default:
+		return 0, false
+	}
+}
+
+//nolint:unused // dead code while we are working on PR 3968
+func (s partialSeriesChunksSet) reparse(idx underfetchedChunksRangeIdx) error {
+	refsRange := s.refsRanges[idx.rangeIdx]
+	ok, err := s.populateRange(idx.rangeIdx, idx.parsed)
+	if err != nil {
+		return fmt.Errorf("parsing underfetched range (block %s first ref %d): %w", idx.blockID, refsRange.firstRef(), err)
+	}
+	if !ok {
+		return fmt.Errorf("chunk length doesn't match after refetching (first ref %d)", refsRange.firstRef())
+	}
+	return nil
+}
+
+// parseRange also corrects the length of the last chunk to the size that it actually is.
+// It slices away any extra bytes in the raw range.
+//
+//nolint:unused // dead code while we are working on PR 3968
+func (s partialSeriesChunksSet) populateRange(rIdx int, dst []storepb.AggrChunk) (bool, error) {
+	rawRange := s.rawRanges[rIdx]
+	r := s.refsRanges[rIdx]
+	for cIdx := range dst {
+		dst[cIdx].MinTime = r.refs[cIdx].minTime
+		dst[cIdx].MaxTime = r.refs[cIdx].maxTime
+		if dst[cIdx].Raw == nil {
+			// This may come as initialized from a pool or from a previous parse. Do an allocation only if it already isn't.
+			dst[cIdx].Raw = &storepb.Chunk{}
+		}
+	}
+	ok, lastChunkLen, totalRead, err := parseRange(rawRange, dst)
+	if err != nil {
+		return false, err
+	}
+	if !ok {
+		// We estimate the length of the last chunk of a series.
+		// Unfortunately, we got it wrong. We can set the length correctly because we now know it.
+		r.refs[len(r.refs)-1].length = lastChunkLen
+		return false, nil
+	}
+	s.rawRanges[rIdx] = s.rawRanges[rIdx][:totalRead]
+	return true, nil
+}
+
+// parseRange parses the byte slice as concatenated encoded chunks. lastChunkLen is non-zero when allChunksComplete==false.
+// An error is returned when gBytes are malformed or when more than the last chunk is incomplete.
+//
+//nolint:unused // dead code while we are working on PR 3968
+func parseRange(rBytes []byte, chunks []storepb.AggrChunk) (allChunksComplete bool, lastChunkLen uint32, totalRead int, _ error) {
```

Comment:
[nit] I would clarify the bool is about whether we successfully read the last chunk.

```suggestion
func parseRange(rBytes []byte, chunks []storepb.AggrChunk) (lastChunkRead bool, lastChunkLen uint32, totalRead int, _ error) {
```